keps/sig-node/20191118-cgroups-v2.md

```diff
@@ -0,0 +1,216 @@
+  - [Option 2: Use cgroups v2 throughout the stack](#option-2-use-cgroups-v2-throughout-the-stack)
+- [Risk and Mitigations](#risk-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+- [Testing](#testing)
+
+<!-- /toc -->
+
+## Summary
+
+A proposal to add support for cgroups v2 to kubernetes.
+
+## Motivation
+
+The new kernel cgroups v2 API was declared stable more than two years
+ago. Newer features in the kernel such as PSI depend upon cgroups
+v2. groups v1 will eventually become obsolete in favor of cgroups v2.
+Some distros are already using cgroups v2 by default, and that
+prevents Kubernetes from working as it is required to run with cgroups
+v1.
+
+## Goals
+
+This proposal aims to:
+
+*   Add support for cgroups v2 to cAdvisor ([https://github.com/google/cadvisor/pull/2309](https://github.com/google/cadvisor/pull/2309))
+*   Add support for cgroups v2 to the Kubelet
+
+## User Stories
+
+*   The Kubelet can run on a host using either cgroups v1 or v2
+
+## Implementation Details
+
+### Dependencies on OCI and container runtimes
+
+Cgroups v2 requires changes to the OCI runtime specs. Changes to the
+OCI specs are out of the scope for this proposal.
+
+### Current status of dependencies
+
+- CRI-O+crun: support cgroups v2
+
+- runc support for cgroups v2 is work in progress
+
+- containerd: [https://github.com/containerd/containerd/pull/3799](https://github.com/containerd/containerd/pull/3799)
+
+- Moby: [https://github.com/moby/moby/pull/40174](https://github.com/moby/moby/pull/40174)
+
+- OCI runtime spec: TODO
+
+## Current cgroups usage and the equivalent in cgroups v2
+
+|Kubernetes cgroups v1|Kubernetes cgroups v2 behavior|
+|---|---|
+|CPU stats for Horizontal Pod Autoscaling|No .percpu cpuacct stats.  Check cadvisor.|
+|CPU pinning based on integral cores|Cpuset controller available|
+|Memory limits|Not changed, different naming|
+|PIDs limits|Not changed, same naming|
+|hugetlb|Added to linux-next, targetting Linux 5.6|
+
+## Proposal
+
+The proposal is to implement cgroups v2 in two different phases.
+
+The first phase ensures that any configuration file designed for
+cgroups v1 will continue to work on cgroups v2.
+
+The second phase requires changes through the entire stack, including
+the OCI runtime specifications.
+
+At startup the Kubelet detects what hierarchy the system is using.  It
+checks the file system type for `/sys/fs/cgroup` (the equivalent of
+`stat -f --format '%T' /sys/fs/cgroup`).  If the type is `cgroup2fs`
+then the Kubelet will use only cgroups v2 during all its execution.
+
+### cgroup namespace
+
+A cgroup namespace restricts the view on the cgroups.  When
+unshare(CLONE_NEWCGROUP) is done, the current cgroup the process
+resides in becomes the root.  Other cgroups won't be visible from the
+new namespace.  It was not enabled by default on a cgroup v1 system as
+older kernel lacked support for it.
+
```

Comment:
Cgroupns is created for pod sandboxes as well as for each of the containers, and the pod sandbox cgroupns is NOT shared by the containers, right?

Reply:
yes right, each container must have their own cgroupns

Reply:
If we create a Pod1 which has container1 and container2.
cgroupns hierarchy will be:
Pod1/container1
Pod1/container2

Then values in Pod1 will be merged values of containers.
It means if each container in Pod1 requests 1GB memory.
memory.limit_in_bytes in Pod1 cgroup will be 2GB as bytes.